src/main/java/ch/njol/skript/effects/EffSort.java

```diff
@@ -0,0 +1,166 @@
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.effects;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Keywords;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.ExprInput;
+import ch.njol.skript.expressions.ExprSortedList;
+import ch.njol.skript.lang.Effect;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.InputSource;
+import ch.njol.skript.lang.ParseContext;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.util.Kleenean;
+import ch.njol.util.Pair;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+@Name("Sort")
+@Description({
+	"Sorts a list variable using either the natural ordering of the contents or the results of the given expression.",
+	"Be warned, this will overwrite the indices of the list variable."
+})
+@Examples({
+	"set {_words::*} to \"pineapple\", \"banana\", \"yoghurt\", and \"apple\"",
+	"sort {_words::*} # alphabetical sort",
+	"sort {_words::*} by length of input # shortest to longest",
+	"sort {_words::*} based on {tastiness::%input%} # sort based on custom value"
+})
+@Since("INSERT VERSION")
+@Keywords("input")
+public class EffSort extends Effect implements InputSource {
+
+	static {
+		Skript.registerEffect(EffSort.class, "sort %~objects% [(by|based on) <.+>]");
+		if (!ParserInstance.isRegistered(InputData.class))
+			ParserInstance.registerData(InputData.class, InputData::new);
+	}
+
+	@Nullable
+	private Expression<?> mappingExpr;
+	@Nullable
+	private String unparsedExpression;
+	private Variable<?> unsortedObjects;
+
+	private Set<ExprInput<?>> dependentInputs = new HashSet<>();
+
+	@Nullable
+	private Object currentValue;
+	@Nullable
+	private String currentIndex;
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (expressions[0].isSingle() || !(expressions[0] instanceof Variable)) {
+			Skript.error("You can only sort list variables!");
```

Comment:
If something accepts the set changer (for the values its returning) do you think it would be safe to allow it here too? I'm 50/50 on this because I know some things don't have analogous setting and getting.

Reply:
My only worry is stuff that returns things of a different format than what it should be set to. I guess we could just hope that if something accepts changes for set with the same class as it returns then it'll be okay.

Reply:
It seems like a bit of a nightmare to properly check the acceptChange classes against the returned class with all the convertedexpression stuff EffChange does, so I'd rather not do this unless I have to.

Reply:
wdym about the convertedexpression stuff?

Reply:
See EffChange's init method